- Added willAddBlock to WorkspaceListener

vicng opened the pull request:
- Updated ProcedureCoordinator to always create definition blocks before the caller blocks, if it doesn't exist
- Changed mutator layouts so they don't automatically perform block bumping after mutations. Callers performing the mutations are now responsible for block bumping.
- Updated FieldLayout to capture change event when setting value from serialized text
- Changed EventManager grouping so it operates as a stack of group events
- Replaced use of NSNotifications with EventManager. Workbench also now fires events when replaying event stack (it previously disabled events)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-ios/333)
<!-- Reviewable:end -->

vicng commented:
+@corydcode +@roboerikg 

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-KfsWj2nS_qzuSyv_HBq:b4ivu6b)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
+@rachel-fenichel to help review the event grouping.

@vicng Event grouping might need a light design doc to explain.

My understanding is you've added a stack of event groups, and when events would be fired from a nested call you add a new copy of the current id to the stack so that when the nested call fires it can safely pop the stack without changing the group. Most things that fire events are using groupAndFireEvents which pairs the calls, but dragging manages it manually.
- Is that a correct summary?
- Roughly how many places do nested event groups happen?
- Which methods are allowed to push and pop group ids directly? Which are allowed to call groupAndFireEvents?

---

Review status: 0 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/UI/Views/MutatorIfElseView.swift, line 184 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwEk-7AebNaIfo4e24:-KfwEk-8H9jVwe-qa-a2:b-7asr10) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/UI/Views/MutatorIfElseView.swift#L184)):*
> ```Swift
>   fileprivate func performMutation() {
>     do {
>       try EventManager.sharedInstance.groupAndFireEvents {
> ```

I think I missed this last time, but why is the view responsible for firing mutator change events? Shouldn't this be a control or model responsibility?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-KfwAyt6veHfbo5Ij3aJ:ba13geo)*
<!-- Sent from Reviewable.io -->

CoryDCode commented:
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Control/ProcedureCoordinator.swift, line 412 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwMOOx0MgGcY5tUDPR:-KfwMOOx0MgGcY5tUDPS:b-ycz8vb) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/Control/ProcedureCoordinator.swift#L412)):*
> ```Swift
> 
>   public func eventManager(_ eventManager: EventManager, didFireEvent event: BlocklyEvent) {
>     if checkProcedureDefinitionNameChangeEvent(event) {
> ```

Can  you add comments for what the point of this is?

---

*[Source/Layout/MutatorIfElseLayout.swift, line 79 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwPkvXeKJ0PMcKudHC:-KfwPkvXeKJ0PMcKudHD:babuce6) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/Layout/MutatorIfElseLayout.swift#L79)):*
> ```Swift
> 
>       // Update UI
>       _ = try layoutCoordinator.rebuildLayoutTree(forBlock: block)
> ```

If this is cool, should we just make rebuildLayoutTree @ discardableResult? Or not return anything, since it looks like you're discarding all of the return values?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333)*
<!-- Sent from Reviewable.io -->

vicng commented:
1) Yep, good summary!
2) Nested event groups only happen during "event chaining" (where listeners make additional changes based on events that have already been fired). As of right now, this only potentially happens for if/return blocks, during a multi-touch drag.
3) Views/controllers are mostly responsible for starting a group, stopping it, and firing events. The only exception is in event chaining, where layouts can start an *existing* group, stop it, and then fire events.

---

Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Control/ProcedureCoordinator.swift, line 412 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwMOOx0MgGcY5tUDPR:-Kfy7J0AdT5mURzShEVg:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/Control/ProcedureCoordinator.swift#L412)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

Can  you add comments for what the point of this is?
</blockquote></details>

Done.

---

*[Source/Layout/MutatorIfElseLayout.swift, line 79 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwPkvXeKJ0PMcKudHC:-Kfy7VHE3nHqKo14z9Lj:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/Layout/MutatorIfElseLayout.swift#L79)):*
<details><summary><i>Previously, CoryDCode wrote…</i></summary><blockquote>

If this is cool, should we just make rebuildLayoutTree @ discardableResult? Or not return anything, since it looks like you're discarding all of the return values?
</blockquote></details>

Done.

---

*[Source/UI/Views/MutatorIfElseView.swift, line 184 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwEk-7AebNaIfo4e24:-Kfy7WT3vq9u3OwVm0fO:bfxhr5u) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/UI/Views/MutatorIfElseView.swift#L184)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

I think I missed this last time, but why is the view responsible for firing mutator change events? Shouldn't this be a control or model responsibility?
</blockquote></details>

The view/controller layer is (mostly) responsible for grouping and firing events.

Adding events are (mostly) done at the layout layer.

Model layer is responsible for nothing :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-Kfy7nzw2zwEFWSTd4Mc:bwjdcub)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
2. Would it be safer to keep a depth count instead of allowing the group to change halfway through a set of events? This would also let you add safety checks to make sure callers aren't trying to change the event group while another event is being generated.

3. I'm uncomfortable with the views having that much responsibility for business logic and that doesn't match your overall design where the layout layer contains the control logic. Is this going to create more work in the future if we do a mobile version of Scratch Blocks?

---

Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/UI/Views/MutatorIfElseView.swift, line 184 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwEk-7AebNaIfo4e24:-Kg0Amr0e_ZbvMN8C57Z:b2auh5m) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/UI/Views/MutatorIfElseView.swift#L184)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

The view/controller layer is (mostly) responsible for grouping and firing events.

Adding events are (mostly) done at the layout layer.

Model layer is responsible for nothing :)


</blockquote></details>

I think you're going to have problems down the road if your view and controls are tightly coupled. We wouldn't have been able to do Scratch Blocks if there hadn't been a rewrite to properly split view and model responsibilities on web a couple years ago.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-Kg0Bc-kWXnNEOdKhFvz:bvwwaux)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
3. Reviewing your design doc. I think I misunderstood the short version, the long version looks good (though let's still discuss it). Sorry I responded too quickly.

---

Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-Kg0E1dZmbJIx_MfdMAB:b-mtkmaa)*
<!-- Sent from Reviewable.io -->

CoryDCode commented:
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Control/ProcedureCoordinator.swift, line 412 at r1](https://reviewable.io:443/reviews/google/blockly-ios/333#-KfwMOOx0MgGcY5tUDPR:-Kg0pNkEApzEwf6Fvve7:b-mszewg) ([raw file](https://github.com/google/blockly-ios/blob/57686fbf2fbfeb927ed10be6133c9bc92c71a5e5/Source/Control/ProcedureCoordinator.swift#L412)):*
<details><summary><i>Previously, vicng wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Discussed offline - we'll change to if (<check>) { process<type>ChangeEvent(...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333)*
<!-- Sent from Reviewable.io -->

vicng commented:
2. After discussing offline, I'm not going to use depth count since it's harder to communicate with the user if they're doing something wrong (ie. do they really want to append to the current group or do they want to start a new group?). Instead, added assertion failures in debug mode (which translate to warnings in release mode) when pushing a new group that doesn't match the existing group. Likewise, added similar assertion failures when user over-pops the group stack.

---

Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-Kg0mdEit0T4ABsQGfut:b1kxdtx)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Control/EventManager.swift, line 158 at r2](https://reviewable.io:443/reviews/google/blockly-ios/333#-Kg1n-JzEE31xXajg-5P:-Kg1n-JzEE31xXajg-5Q:b-k3m9wa) ([raw file](https://github.com/google/blockly-ios/blob/cd7d7f76571947e1049b91f04f75f2a71b748894/Source/Control/EventManager.swift#L158)):*
> ```Swift
>    - parameter groupID: The groupID to push.
>    - note: It is not recommended to push a group ID that differs from `self.currentGroupID`,
>    if it is not `nil`. Doing this will result in an error in debug mode and a warning in release mode.
> ```

nit line length

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333#-:-Kg1nuSrTq3H4v8V7uwI:b-uwkowd)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[Source/Control/EventManager.swift, line 158 at r2](https://reviewable.io:443/reviews/google/blockly-ios/333#-Kg1n-JzEE31xXajg-5P:-Kg1oQtvvIYlvqYlFYmp:b-896fix) ([raw file](https://github.com/google/blockly-ios/blob/cd7d7f76571947e1049b91f04f75f2a71b748894/Source/Control/EventManager.swift#L158)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

nit line length
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-ios/333)*
<!-- Sent from Reviewable.io -->
